contracts/nois-gateway/src/contract.rs

```diff
@@ -110,14 +131,58 @@ pub fn ibc_channel_open(
 
 #[entry_point]
 pub fn ibc_channel_connect(
-    _deps: DepsMut,
-    _env: Env,
+    deps: DepsMut,
+    env: Env,
     msg: IbcChannelConnectMsg,
 ) -> StdResult<IbcBasicResponse> {
     let channel = msg.channel();
     let chan_id = &channel.endpoint.channel_id;
 
+    let config = CONFIG.load(deps.storage)?;
+
+    // Instantiate payment contract
+    let creator = deps.api.addr_canonicalize(env.contract.address.as_str())?;
+    let CodeInfoResponse { checksum, .. } =
+        deps.querier.query_wasm_code_info(config.payment_code_id)?;
+    let salt = hash_channel(chan_id);
+    #[allow(unused)]
+    let address = instantiate2_address(&checksum, &creator, &salt)
+        .map_err(|e| StdError::generic_err(format!("Cound not generate address: {}", e)))?;
+
+    let address = {
+        #[cfg(not(test))]
+        {
+            deps.api.addr_humanize(&address)?
+        }
+        #[cfg(test)]
+        cosmwasm_std::Addr::unchecked("some paymane address")
+    };
+    PAYMENT_ADDRESSES.save(deps.storage, chan_id, &address)?;
+
+    let msg = WasmMsg::Instantiate2 {
+        admin: Some(env.contract.address.into()), // Only gateway can update the contracts it created
+        code_id: config.payment_code_id,
+        label: format!("For {chan_id}"),
+        msg: to_binary(&nois_payment::msg::InstantiateMsg {
+            sink: config.sink.into(),
+            community_pool: config.community_pool.into(),
+        })?,
+        funds: vec![],
+        salt,
+    };
+
+    // Send info to proxy
+    let packet = IbcMsg::SendPacket {
+        channel_id: chan_id.to_string(),
+        data: to_binary(&OutPacket::Welcome {
+            payment: address.into(),
+        })?,
+        timeout: env.block.time.plus_seconds(WELCOME_PACKET_LIFETIME).into(),
+    };
```

Comment:
I think we need to ensure the direction of the connection is correct and this is called last (Nois is chain B, ChanOpenConfirm is called here).

Otherwise we try to send a packet over a channel in `(OPEN, TRYOPEN)` state. Not sure if this is ready for sending packets.
https://github.com/cosmos/ibc/tree/main/spec/core/ics-004-channel-and-packet-semantics#channel-lifecycle-management

Reply:
Done in 030fc27bd6844b49ea0f4633f29387e8fc4e1033